airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/message/MultiProducerChannel.kt

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.message
+
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.lang.IllegalStateException
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicLong
+import kotlinx.coroutines.channels.Channel
+
+/**
+ * A channel designed for use with a dynamic amount of producers. Close will only close the
+ * underlying channel, when there are no remaining registered producers.
+ */
+class MultiProducerChannel<T>(override val channel: Channel<T>) : ChannelMessageQueue<T>() {
```

Comment:
If this pattern repeats, we could upgrade this into something that wraps any closable.